matrix_sdk_base/src/client.rs

```diff
@@ -1405,42 +1420,58 @@ impl BaseClient {
             }
         };
 
-        match event {
-            RoomEvent::RoomMember(mem) => event_emitter.on_room_member(room, &mem).await,
-            RoomEvent::RoomName(name) => event_emitter.on_room_name(room, &name).await,
-            RoomEvent::RoomCanonicalAlias(canonical) => {
-                event_emitter
-                    .on_room_canonical_alias(room, &canonical)
-                    .await
-            }
-            RoomEvent::RoomAliases(aliases) => event_emitter.on_room_aliases(room, &aliases).await,
-            RoomEvent::RoomAvatar(avatar) => event_emitter.on_room_avatar(room, &avatar).await,
-            RoomEvent::RoomMessage(msg) => event_emitter.on_room_message(room, &msg).await,
-            RoomEvent::RoomMessageFeedback(msg_feedback) => {
-                event_emitter
-                    .on_room_message_feedback(room, &msg_feedback)
-                    .await
-            }
-            RoomEvent::RoomRedaction(redaction) => {
-                event_emitter.on_room_redaction(room, &redaction).await
-            }
-            RoomEvent::RoomPowerLevels(power) => {
-                event_emitter.on_room_power_levels(room, &power).await
-            }
-            RoomEvent::RoomTombstone(tomb) => event_emitter.on_room_tombstone(room, &tomb).await,
-            RoomEvent::CustomRoom(custom) => {
-                event_emitter
-                    .on_unrecognized_event(room, &CustomOrRawEvent::CustomRoom(custom))
-                    .await
-            }
-            _ => {}
+        match &event {
+            AnyRoomEventStub::State(event) => match &event {
+                AnyStateEventStub::RoomMember(e) => event_emitter.on_room_member(room, &e).await,
+                AnyStateEventStub::RoomName(e) => event_emitter.on_room_name(room, &e).await,
+                AnyStateEventStub::RoomCanonicalAlias(e) => {
+                    event_emitter.on_room_canonical_alias(room, &e).await
+                }
+                AnyStateEventStub::RoomAliases(e) => event_emitter.on_room_aliases(room, &e).await,
+                AnyStateEventStub::RoomAvatar(e) => event_emitter.on_room_avatar(room, &e).await,
+                AnyStateEventStub::RoomPowerLevels(e) => {
+                    event_emitter.on_room_power_levels(room, &e).await
+                }
+                AnyStateEventStub::RoomTombstone(e) => {
+                    event_emitter.on_room_tombstone(room, &e).await
+                }
+                AnyStateEventStub::RoomJoinRules(_e) => {
+                    // TODO is this needed ??
```

Comment:
I would say it is, consider a client like Riot. Riot changes the decoration based on the join rule, if the room is invite only it gets decorated with a lock.